File: javascript/test/instance/GenericCollectionHolder_FindFirstIndexedOrNullAlias.ts
//··························································
// Copyright (c) 2023-2025. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {NullOr}                                                   from "@joookiwi/type"
import type {ReverseBooleanCallback, ReverseRestrainedBooleanCallback} from "../../src/type/callback"
import type {StraightCollectionHolderForTest}                          from "./StraightCollectionHolderForTest"

import {GenericCollectionHolder} from "../../src/GenericCollectionHolder"
import {ABCD}                    from "../value/arrays"

export class GenericCollectionHolder_FindFirstIndexedOrNullAlias
    extends GenericCollectionHolder<string, readonly string[]>
    implements StraightCollectionHolderForTest<string> {

    public amountOfCall = 0

    public constructor() { super(ABCD,) }

    public execute(action: (instance: this,) => void,): this {
        action(this,)
        return this
    }

    public override findFirstIndexedOrNull<const S extends string, >(predicate: ReverseRestrainedBooleanCallback<string, S>,): NullOr<S>
    public override findFirstIndexedOrNull(predicate: ReverseBooleanCallback<string>,): NullOr<string>
    public override findFirstIndexedOrNull(predicate: ReverseBooleanCallback<string>,) {
        this.amountOfCall++
        return super.findFirstIndexedOrNull(predicate,)
    }

}

File: java/collection.joookiwi/src/test/java/instance/GenericCollectionHolder_FindFirstIndexedOrNullAlias.java
package instance;

import java.util.function.BooleanSupplier;
import java.util.function.Consumer;
import java.util.function.IntPredicate;
import joookiwi.collection.java.GenericCollectionHolder;
import joookiwi.collection.java.callback.IntObjPredicate;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_THIS_1;
import static joookiwi.collection.java.NumericConstants.MAX_INT_VALUE;
import static value.Arrays.ABCD;

@NotNullByDefault
public final class GenericCollectionHolder_FindFirstIndexedOrNullAlias
        extends GenericCollectionHolder<String>
        implements StraightCollectionHolderForTest<String, GenericCollectionHolder_FindFirstIndexedOrNullAlias> {

    /// The field that tell the amount of time the method `findFirstIndexedOrNull` has been called
    public @Range(from = 0, to = MAX_INT_VALUE) int amountOfCall = 0;

    public GenericCollectionHolder_FindFirstIndexedOrNullAlias() { super(ABCD); }

    @Override public @Range(from = 0, to = MAX_INT_VALUE) int getAmountOfCall() { return amountOfCall; }

    @Contract(ALWAYS_THIS_1)
    @Override public GenericCollectionHolder_FindFirstIndexedOrNullAlias execute(final Consumer<GenericCollectionHolder_FindFirstIndexedOrNullAlias> action) {
        action.accept(this);
        return this;
    }


    @Override public @Nullable String findFirstIndexedOrNull(final IntObjPredicate<? super String> predicate) {
        amountOfCall++;
        return super.findFirstIndexedOrNull(predicate);
    }

    @Override public @Nullable String findFirstIndexedOrNull(final IntPredicate predicate) {
        amountOfCall++;
        return super.findFirstIndexedOrNull(predicate);
    }

    @Override public @Nullable String findFirstIndexedOrNull(final BooleanSupplier predicate) {
        amountOfCall++;
        return super.findFirstIndexedOrNull(predicate);
    }

}
